assets/other-scripts/recaptcha/index.js

```diff
@@ -34,89 +24,180 @@ function domReady( callback ) {
 
-				const widgetId = grecaptcha.render( submitButton || container, options );
-				newspack_grecaptcha.widgets[ containerId ] = widgetId;
-			} );
-		}
-	} );
+		// Refresh reCAPTCHAs on Woo checkout update and error.
+		( function ( $ ) {
+			if ( ! $ ) {
+				return;
+			}
+			$( document ).on( 'updated_checkout', () => refresh( field, action ) );
+			$( document.body ).on( 'checkout_error', () => refresh( field, action ) );
+		} )( jQuery );
+	}
 }
 
 /**
- * Refresh all reCAPTCHA tokens.
+ * Remove the hidden reCAPTCHA v3 token field from the given form.
+ *
+ * @param {HTMLElement} form The form element.
  */
-function refreshCaptchas() {
-	if ( isV2 ) {
-		const { widgets } = newspack_grecaptcha;
-		for ( const containerId in widgets ) {
-			grecaptcha.reset( newspack_grecaptcha.widgets[ containerId ] );
-		}
+function removeHiddenField( form ) {
+	const field = form.querySelector( 'input[name="g-recaptcha-response"]' );
+	if ( field ) {
+		field.parentElement.removeChild( field );
 	}
 }
 
 /**
- * Fetch a reCAPTCHA token via the v3 JS API. Only needed for reCAPTCHA v3.
- * v2 automatically generates a token on form submission, so if using v2 the
- * promise can be silently resolved.
+ * Refresh the reCAPTCHA v2 widget attached to the given element.
  *
- * See: https://developers.google.com/recaptcha/docs/v3#programmatically_invoke_the_challenge
+ * @param {HTMLElement} el Element with the reCAPTCHA widget to refresh.
+ */
+function refreshWidget( el ) {
+	const widgetId = parseInt( el.getAttribute( 'data-recaptcha-widget-id' ) );
+	if ( ! isNaN( widgetId ) ) {
+		grecaptcha.reset( widgetId );
+	}
+}
+
+/**
+ * Render reCAPTCHA v2 widget on the given form.
  *
- * @return {Promise<string>} The reCAPTCHA token, if needed, or an empty string.
+ * @param {HTMLElement} form The form element.
  */
-function getCaptchaV3Token() {
-	return new Promise( ( res, rej ) => {
-		if ( ! grecaptcha || ! isV3 ) {
-			return res( '' );
+function renderWidget( form ) {
+	const submitButtons = [
+		...form.querySelectorAll( 'input[type="submit"], button[type="submit"]' ),
+	];
+	const options = {
+		sitekey: siteKey,
+		size: isInvisible ? 'invisible' : 'normal',
+	};
+
+	submitButtons.forEach( button => {
+		// Don't render widget if the button has a data-skip-recaptcha attribute.
+		if ( button.hasAttribute( 'data-skip-recaptcha' ) ) {
+			return;
 		}
 
-		if ( ! grecaptcha?.ready ) {
-			rej( 'Error loading the reCAPTCHA library.' );
+		// Refresh widget if it already exists.
+		if ( button.hasAttribute( 'data-recaptcha-widget-id' ) ) {
+			refreshWidget( button );
+			return;
 		}
 
-		grecaptcha.ready( () => {
-			grecaptcha
-				.execute( siteKey, { action: 'submit' } )
-				.then( token => res( token ) )
-				.catch( e => rej( e ) );
+		// Callback when reCAPTCHA passes validation.
+		const callback = () => {
+			form.requestSubmit( button );
+			refreshWidget( button );
+		};
+		const widgetId = grecaptcha.render( button, {
```

Comment:
Might be a good idea to re-focus the form for the reader to ensure recaptcha is visible on longer forms like modal checkout.

Reply:
It seems like the reCAPTCHA widget styles aren't playing nice with some styles in our modal checkout iframe. The widget is supposed to be fixed to the top of the viewport, so it shouldn't ever be rendering outside of the visible scroll. I spent some time trying to figure out why, but couldn't easily, so 62c60e87 adds a scroll to the top when the widget is rendered. I also don't know why `window.scroll` wasn't working here, hence the use of `document.body.scrollIntoView`.